src/main/java/com/google/sps/servlets/DeleteFolderServlet.java

```diff
@@ -0,0 +1,175 @@
+
+import static com.google.appengine.api.datastore.TransactionOptions.Builder.*;
+
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.api.datastore.Transaction;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.KeyFactory;
+
+import com.google.appengine.api.taskqueue.Queue;
+import com.google.appengine.api.taskqueue.QueueFactory;
+import com.google.appengine.api.taskqueue.TaskOptions;
+
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreFailureException;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import com.google.gson.Gson;
+import com.google.sps.tool.ResponseSerializer;
+
+@WebServlet("/deletefolder")
+public class DeleteFolderServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+
+    if (userService.isUserLoggedIn()) {
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      String folderKey = request.getParameter("folderKey");
+      Entity folder = getExistingFolderInDatastore(datastore, folderKey);
+
+      if (folder == null) {
+        String jsonErrorInfo =
+            ResponseSerializer.getErrorJson("Cannot delete Folder at the moment");
+        response.setContentType("application/json;");
+        response.getWriter().println(new Gson().toJson(jsonErrorInfo));
+      } else {
+        deleteAllCardsInsideFolder(folder);
+        deleteFolder(folder);
+      }
+    }
+  }
+
+  private Entity getExistingFolderInDatastore(DatastoreService datastore, String cardKey)
+      throws IOException {
+    try {
+      return datastore.get(KeyFactory.stringToKey(cardKey));
+    } catch (EntityNotFoundException e) {
+      return null;
+    }
+  }
+
+  private void deleteFolder(Entity folder) {
+    int retries = 5;
+    while (true) {
+      try {
+        folderDeletionTransaction(folder);
+        break;
+      } catch (Exception e) {
+        if (retries == 0) {
+          addDatastoreTaskToQueue(folder);
+          break;
+        }
+        --retries;
+      }
+    }
+  }
+
+  private void folderDeletionTransaction(Entity folder) throws IOException {
```

Comment:
The name of this object is a noun, which makes me think it returns an object. But it's actually a procedure, so it should be a short phrase (`deleteFolderImpl`? Or, `deleteFolder -> deleteFolderWithRetries` and rename this method `deleteFolder`.)

The transaction is unnecessary because you're only deleting one thing. Transactions are for making multiple operations [atomic](https://en.wikipedia.org/wiki/Atomicity_(database_systems)); if you're only executing a single operation, it's already atomic, so the transaction is redundant.

Reply:
Thanks Tyler. The concept of atomicity is making sense.